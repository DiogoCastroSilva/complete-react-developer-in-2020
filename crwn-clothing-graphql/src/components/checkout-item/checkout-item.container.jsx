import React from 'react';
import { gql } from 'apollo-boost';
import { Mutation } from 'react-apollo';
import CheckoutItemComponent from './checkout-item.component';



const ADD_ITEM_TO_CART = gql`
    mutation AddItemToCart($item: Item!) {
        addItemToCart(item: $item) @client
    }
`;

const REMOVE_ITEM_FROM_CART = gql`
  mutation RemoveItemFromCart($item: Item!) {
    removeItemFromCart(item: $item) @client
  }
`;

const CLEAR_ITEM_FROM_CART = gql`
  mutation ClearItemFromCart($item: Item!) {
    clearItemFromCart(item: $item) @client
  }
`;

const CheckoutItemContainer = (props) => (
    <Mutation mutation={ADD_ITEM_TO_CART}>
        {addItemToCart =>
            <Mutation mutation={REMOVE_ITEM_FROM_CART}>
                {removeItemFromCart =>
                    <Mutation mutation={CLEAR_ITEM_FROM_CART}>
                        {clearItemFromCart =>
                            <CheckoutItemComponent
                                addItem={item => addItemToCart({ variables: { item } })}
                                removeItem={item => removeItemFromCart({ variables: { item } })}
                                clearItem={item => clearItemFromCart({ variables: { item } })}
                                {...props}
                            />
                        }
                    </Mutation>
                }
            </Mutation>
        }
    </Mutation>
);

export default CheckoutItemContainer;