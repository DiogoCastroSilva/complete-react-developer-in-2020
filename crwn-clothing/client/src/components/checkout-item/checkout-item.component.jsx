import React from 'react';

import {
    CheckoutItemContainer,
    ImageContainer,
    TextContainer,
    QuantityContainer,
    RemoveButtonContainer
} from './checkout-item.styles';


const CheckoutItem = ({ item, removeItem, addItem, clearItem }) => {
    const { name, quantity, price, imageUrl } = item;
    return (
        <CheckoutItemContainer>
            <ImageContainer>
                <img src={imageUrl} alt='item' />
            </ImageContainer>
            <TextContainer>{name}</TextContainer>
            <QuantityContainer>
                <div
                    onClick={() => removeItem(item)}
                >
                    &#10094;
                </div>
                <span>{quantity}</span>
                <div
                    onClick={() => addItem(item)}
                >
                    &#10095;
                </div>
            </QuantityContainer>
            <TextContainer>{price}</TextContainer>
            <RemoveButtonContainer
                className='remove-button'
                onClick={() => clearItem(item)}
            >
                &#10005;
            </RemoveButtonContainer>
        </CheckoutItemContainer>
    );
};

export default CheckoutItem;