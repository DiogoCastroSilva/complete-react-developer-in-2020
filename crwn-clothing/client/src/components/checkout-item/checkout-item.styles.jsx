import styled from 'styled-components';



export const CheckoutItemContainer = styled.div`
    width: 100%;
    display: flex;
    min-height: 100px;
    border-bottom: 1px solid darkgrey;
    padding: 15px 0;
    font-size: 20px;
    align-items: center;

    @media screen and (max-width: 800px) {
        font-size: 18px;
    }
`;

export const ImageContainer = styled.div`
    width: 23%;
    padding-right: 15px;
  
    img {
        width: 100%;
        height: 100%;
    }
`;

export const TextContainer = styled.span`
    width: 23%;

    @media screen and (max-width: 800px) {
        width: 22%;
    }
`;

export const QuantityContainer = styled(TextContainer)`
    display: flex;

    div {
        cursor: pointer;
    }

    span {
        margin: 0 10px;
    }
`;

export const RemoveButtonContainer = styled.span`
    padding-left: 12px;
      cursor: pointer;
`;