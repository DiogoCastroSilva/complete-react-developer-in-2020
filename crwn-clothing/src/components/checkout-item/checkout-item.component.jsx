import React from 'react';
import { connect } from 'react-redux';

import './checkout-item.styles.scss';

import { clearItem, removeItem, addItem } from '../../redux/cart/cart.actions';


const CheckoutItem = ({ item, removeItem, addItem, clearItem }) => {
    const { name, quantity, price, imageUrl } = item;
    return (
        <div className='checkout-item'>
            <div className='image-container'>
                <img src={imageUrl} alt='item' />
            </div>
            <span className='name'>{name}</span>
            <span className='quantity'>
                <div
                    onClick={() => removeItem(item)}
                    className='arrow'
                >
                    &#10094;
                </div>
                <span className='value' >
                    {quantity}
                </span>
                <div
                    onClick={() => addItem(item)}
                    className='arrow'
                >
                    &#10095;
                </div>
            </span>
            <span className='price'>{price}</span>
            <span
                className='remove-button'
                onClick={() => clearItem(item)}
            >
                &#10005;
            </span>
        </div>
    );
};

const matDispatchToProps = dispatch => {
    return {
        removeItem: item => dispatch(removeItem(item)),
        addItem: item => dispatch(addItem(item)),
        clearItem: item => dispatch(clearItem(item))
    }
};

export default connect(null, matDispatchToProps)(CheckoutItem);