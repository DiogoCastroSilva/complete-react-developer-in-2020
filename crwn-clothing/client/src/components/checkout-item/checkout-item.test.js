import React from 'react';
import { shallow } from 'enzyme';

import CheckoutItem from './checkout-item.component';
import { RemoveButtonContainer, QuantityContainer } from './checkout-item.styles';


describe('Testing CheckoutItem component', () => {
    let wrapper;
    let mockRemoveItem;
    let mockAddItem;
    let mockClearItem;

    beforeEach(() => {
        mockRemoveItem = jest.fn();
        mockAddItem = jest.fn();
        mockClearItem = jest.fn();

        const mockProps = {
            item: {
                imageUrl: 'https://store.storeimages.cdn-apple.com/4668/as-images.apple.com/is/iphone-12-pro-family-hero?wid=940&amp;hei=1112&amp;fmt=jpeg&amp;qlt=80&amp;op_usm=0.5,0.5&amp;.v=1604021663000',
                price: '1179â‚¬',
                name: 'Iphone 12 Pro',
                quantity: 1
            },
            clearItem: mockClearItem,
            addItem: mockAddItem,
            removeItem: mockRemoveItem
        };

        wrapper = shallow(<CheckoutItem {...mockProps} />);
    });

    it('expects to render the CheckoutItem component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('expects to call the AddItem function', () => {
        wrapper
            .find(QuantityContainer)
            .childAt(2)
            .simulate('click');
        
        expect(mockAddItem).toHaveBeenCalled();
    });

    it('expects to call the RemoveItem function', () => {
        wrapper
            .find(QuantityContainer)
            .childAt(0)
            .simulate('click');

        expect(mockRemoveItem).toHaveBeenCalled();
    });

    it('expects to call the ClearItem function', () => {
        wrapper.find(RemoveButtonContainer).simulate('click');
        expect(mockClearItem).toHaveBeenCalled();
    });
});