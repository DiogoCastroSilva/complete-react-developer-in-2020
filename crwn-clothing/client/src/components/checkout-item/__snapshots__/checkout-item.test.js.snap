// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing CheckoutItem component expects to render the CheckoutItem component 1`] = `
<styled.div>
  <styled.div>
    <img
      alt="item"
      src="https://store.storeimages.cdn-apple.com/4668/as-images.apple.com/is/iphone-12-pro-family-hero?wid=940&amp;hei=1112&amp;fmt=jpeg&amp;qlt=80&amp;op_usm=0.5,0.5&amp;.v=1604021663000"
    />
  </styled.div>
  <styled.span>
    Iphone 12 Pro
  </styled.span>
  <Styled(styled.span)>
    <div
      onClick={[Function]}
    >
      ❮
    </div>
    <span>
      1
    </span>
    <div
      onClick={[Function]}
    >
      ❯
    </div>
  </Styled(styled.span)>
  <styled.span>
    1179€
  </styled.span>
  <styled.span
    className="remove-button"
    onClick={[Function]}
  >
    ✕
  </styled.span>
</styled.div>
`;
